File: ros2_ws_p/src/number_publisher/number_publisher/number_counter.py
import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int64

class NumberCounter(Node):
    def __init__(self):
        super().__init__("number_counter")
        self.counter_ = 0
        self.number_subscriber_ = self.create_subscription(
            Int64,
            "/number",
            self.callback_number,
            10
        )
        self.get_logger().info("[NumberCounter] Node has started.")

    def callback_number(self, msg: Int64):
        self.counter_ += msg.data
        self.get_logger().info("Counter: " + str(self.counter_))    
    
def main(args = None):
    rclpy.init(args = args)
    node = NumberCounter()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: ros2_ws_p/src/number_publisher_cpp/src/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"
#include <functional>

using namespace std::placeholders;

class NumberCounter : public rclcpp::Node
{
    public:
        NumberCounter() : Node("number_counter")
        {
            number_subscriber_ = this->create_subscription<example_interfaces::msg::Int64>(
            "/number", 10, std::bind(&NumberCounter::callbackNumber, this, _1));

            RCLCPP_INFO(this->get_logger(), "[NumberCounter] Node has been started.");
        }
        

    private:
        rclcpp::Subscription<example_interfaces::msg::Int64>::SharedPtr number_subscriber_;
        int64_t counter_ = 0;

        void callbackNumber(const example_interfaces::msg::Int64::SharedPtr msg)
        {
           counter_ += msg->data;
           RCLCPP_INFO(this->get_logger(), "Counter: %ld", counter_);
        }
};

int main(int argc, char** argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberCounter>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    RCLCPP_INFO(node->get_logger(), "[NumberCounter] Node has been shut down.");

    return 0;
}